cli/tre/authentication.py

```diff
@@ -0,0 +1,48 @@
+import asyncio
+from logging import Logger
+import msal
+from azure.identity.aio import ClientSecretCredential
+from azure.cli.core import cloud
+from msal.authority import AuthorityBuilder
+
+
+def get_auth_token_client_credentials(
+    log: Logger,
+    client_id: str,
+    client_secret: str,
+    aad_tenant_id: str,
+    api_scope: str,
+    verify: bool
+):
+    try:
+        event_loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(event_loop)
+
+        credential = ClientSecretCredential(aad_tenant_id, client_id, client_secret, connection_verify=verify, authority=get_aad_authority_fqdn())
+        token = event_loop.run_until_complete(credential.get_token(f'{api_scope}/.default'))
+        event_loop.run_until_complete(credential.close())
+
+        event_loop.close()
+        return token.token
+    except Exception as ex:
+        log.error(f"Failed to authenticate: {ex}")
+        raise RuntimeError("Failed to get auth token")
+
+
+def get_public_client_application(
+    client_id: str,
+    aad_tenant_id: str,
+    token_cache
+):
+    return msal.PublicClientApplication(
+        client_id=client_id,
+        authority=AuthorityBuilder(instance=get_aad_authority_fqdn(), tenant=aad_tenant_id),
+        token_cache=token_cache)
+
+
+def get_cloud() -> cloud.Cloud:
+    return cloud.get_active_cloud()
+
+
+def get_aad_authority_fqdn() -> str:
+    return get_cloud().endpoints.active_directory.replace('https://', '')
```

Comment:
nit: it's generally better to use `urllib` to parse and manipulate URLs

Reply:
also, this is repeated code
perhaps we need a TRE utils package that we can import into multiple things

Reply:
Repeated code is a repo-wide issue

If you think this specific case is fixable in this PR then great, but feel free to resolve this as "won't fix". It's fine as-is.